src/modeler/loadFiles/checkLocation.cpp

```diff
@@ -0,0 +1,200 @@
+                      + "': Error for variable '" + varNode->value()
+                      + "': reduced_cost can only be used on variables located in subproblems");
+                }
+            }
+        }
+
+        if (functionNode->type() == FunctionNodeType::dual)
+        {
+            // This node contains the constraint name
+            const auto* n = dynamic_cast<const ParameterNode*>(functionNode->getOperands().at(0));
+            for (const auto& constraint: model.Constraints())
+            {
+                if (constraint.Id() == n->value() && constraint.location() != Location::SUBPROBLEMS)
+                {
+                    throw LocationError(
+                      "Model '" + model.Id() + "': In expression '" + exprStr
+                      + "': Error for constraint '" + n->value()
+                      + "': dual can only be used on constraints located in subproblems");
+                }
+            }
+        }
+    }
+}
+
+void checkExpression(const Node* expression,
+                     const Location& location,
+                     Model& model,
+                     const std::string& exprStr,              // used for error msgs
+                     const std::string& errorMsgForPortField) // used for error msgs
+{
+    for (const auto& node: ASTconst(expression))
+    {
+        // base variable
+        if (const auto* varNode = dynamic_cast<const VariableNode*>(&node); varNode)
+        {
+            for (const auto& variable: model.Variables())
+            {
+                if (variable.Id() == varNode->value()
+                    && !AreLocationsCompatibleForExpressions(variable.location(), location))
+                {
+                    throw LocationError(
+                      errorMsgForPortField + "Model '" + model.Id() + "': In expression '" + exprStr
+                      + "': Error for variable '" + varNode->value()
+                      + "': Location doesn't match the expression location (variable location: "
+                      + LocationToStr(variable.location())
+                      + ", expression location: " + LocationToStr(location) + ")");
+                }
+            }
+            continue;
+        }
+
+        // Portfields can contains variables, we recursively check their expressions
+        if (const auto* portFieldNode = dynamic_cast<const PortFieldNode*>(&node); portFieldNode)
+        {
+            // This code allows to have a clear message if one of the referenced expression contains
+            // bad locations
+            std::string msgInCaseOfError = "In model '" + model.Id() + "': In expression '"
+                                           + exprStr + "': this port field definition '"
+                                           + portFieldNode->getPortName() + "."
+                                           + portFieldNode->getFieldName()
+                                           + "': is referencing a expression with a bad location";
```

Comment:
Display the location and accepted values
```suggestion
                                           + "': is referencing a expression with an unsupported location";
```

Reply:
I used incorrect, I feel like unsupported means that we don't have implemented something on our side